src/rest/openapi.yaml

```
@@ -0,0 +1,455 @@
+        pending dataset if you would like to change the current active opertional dataset if the Thread node is active.
+
+        Default parameters are choosen for parameters not set in the request body.
+      requestBody:
+        description: |-
+          Operational dataset object that will be stored as active operational dataset. JSON keys which are not set will be initialized with
+          default or random values.
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/Dataset"
+      responses:
+        "202":
+          description: Successfully created the active operational dataset.
+        "400":
+          description: Invalid request body.
+    put:
+      tags:
+        - node
+      summary: Update current active opertional dataset
+      description: |-
+        Updates the current active operational dataset on the current node. Only allowed if the Thread node is inactive. Create a
+        pending dataset if you would like to change the current active opertional dataset if the Thread node is active.
+      requestBody:
+        description: |-
+          Operational dataset object that will be stored as active operational dataset. Supports request body Content-Type `text/plain`
+          (dataset in TLV format as hex string) or `application/json` (dataset in JSOn format, JSON keys which are not set will be
+          initialized with default or random values).
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/Dataset"
+          plain/text:
+            schema:
+              $ref: "#/components/schemas/TlvDataset"
+      responses:
+        "202":
+          description: Successfully updated the active operational dataset.
+        "400":
+          description: Invalid request body.
+        "404":
+          description: No active opertional dataset to update.
+  /node/dataset/pending:
+    get:
+      tags:
+        - node
+      summary: Get current pending operational dataset
+      responses:
+        "200":
+          description: Returns currently pending opertional dataset
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Dataset"
+            text/plain:
+              schema:
+                $ref: "#/components/schemas/TlvDataset"
+        "204":
+          description: No pending operational dataset
+    post:
+      tags:
+        - node
+      summary: Create new pending opertional dataset
+      description: |-
+        Creates new pending operational dataset on the current node. If the delay timer is specified, the new pending operational dataset
+        will become active after the time elapsed. The pending dataset will become empty.
+
+        Default parameters are choosen for parameters not set in the request body. By default, the delay timer is *not* set.
```

Does it mean the delay timer **MUST** be included in the the request body?

---

No, but if its not set the pending operational dataset will just not get applied. Maybe extend that?

```suggestion
        Default parameters are chosen for parameters not set in the request body. By default, the delay timer is *not* set, which means
        the pending dataset will get stored but not automatically applied as active dataset.
```

---

No, but if its not set the pending operational dataset will just not get applied. Maybe extend that?

```suggestion
        Default parameters are chosen for parameters not set in the request body. By default, the delay timer is *not* set, which means
        the pending dataset will get stored but not automatically applied as active dataset.
```

---

So far I avoided adding too much "business logic" to the REST API, adding an implicit default for the REST API only feels wrong to me. However, I am fine enforcing a delay timer value.

---

> However, I am fine enforcing that a delay timer is present.

This works for me

---

I've discovered one downside with that approach when using TLVs: The client needs to be able to add a new field to the TLV.

Before, we could set a pending operational dataset using TLV  (e.g. from Android API), and then set the delay timer using a JSON request.

Now, the client needs to be able to extend the TLV we get from Android API and extend that with a delay timer :cold_sweat: 

Adding a default timer would be a work around. But I actually kinda liked the behavior before where we were able to modify the pending dataset and then set the delay timer in one last request.

---

> Before, we could set a pending operational dataset using TLV (e.g. from Android API), and then set the delay timer using a JSON request.
>
> Now, the client needs to be able to parse the TLV we get from Android API and extend that with a delay timer cold_sweat

What's the "Android API" you are referring to? a Pending Operational Dataset without a Delay Timer shouldn't be considered valid I believe.

---

Is it https://developers.home.google.com/thread?

---

> What's the "Android API" you are referring to? a Pending Operational Dataset without a Delay Timer shouldn't be considered valid I believe.

Yes, the API you've linked. Well, if it is an active operational dataset, no delay timer is normal right?

Maybe it is possible to get a dataset and modify it using the Android API, I am not very familiar with it.

Let's assume you'd use this REST API to synchronize the credentials of two OTBR's. One of the OTBR is already running, so you are reading its active operational dataset. Now you need to add a timer delay to it, and set it as pending. You have to have the knowledge how to change the TLVs...

---

> Yes, the API you've linked. Well, if it is an active operational dataset, no delay timer is normal right?

Yes, Active Operational Dataset doesn't have delay timer.

One path is we add new Android/GMS API to convert Active Operational Dataset to Pending Dataset, but this will generally take a few month to release.

Another solution (and I prefer) is to define Pending Operational Dataset as a composition of `Active Operational Dataset`, `Pending Timestamp` and `Delay Timer`. For example:

```
// PendingOperational Dataset
{
    "ActiveDataset": "000102030ADEFBCDD...",
    "PendingTimestamp": "1",
    "DelayTimer": "200",
}

// And format with JSON ActiveDataset
{
    "ActiveDataset": {
        "Channel": "13",
        "NetworkName": "MyThreadNet",
        // ......
    },
    "PendingTimestamp": "1",
    "DelayTimer": "200",
}

```

We don't need to define strict 1-to-1 mapping to the openthread API for REST - I think we can define the REST API in a way which it's most easy-to-use for typical REST clients such as Apps or service backends. Those client don't need to know the underlying openthread API or structs, so we are free to define the REST data structure.

---

The REST Operational Dataset definition is even better because it expresses the relationship that "A Pending Dataset includes an Active Dataset"

---

I've implemented what we discussed yesterday (`PUT` only, same semantics for JSON and TLV). It requires a change in OpenThread as well: https://github.com/openthread/openthread/pull/8669

I plan to tackle Pending Operational Dataset as a composition of Active Operational Dataset, Pending Timestamp and Delay Timer as well.